cli/data.go

```diff
@@ -35,89 +37,182 @@ func (c *AppClient) BinaryData(dst string, filter *datapb.Filter) error {
+	wg := sync.WaitGroup{}
+
+	// In one routine, get all IDs matching the filter and pass them into ids.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		var limit int
+		if concurrentDownloads > 100 {
+			limit = 100
+		} else {
+			limit = concurrentDownloads
+		}
+		if err := getMatchingBinaryIDs(ctx, c.dataClient, filter, ids, limit); err != nil {
+			errs <- err
+			cancel()
+		}
+	}()
+
+	// In parallel, read from ids and download the binary for each id in batches of defaultConcurrentDownloads.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		var nextID string
+		var done bool
+		numFilesDownloaded := &atomic.Int32{}
+		downloadWG := sync.WaitGroup{}
+		for {
+			for i := 0; i < concurrentDownloads; i++ {
+				if err := ctx.Err(); err != nil {
+					errs <- err
+					cancel()
+					done = true
+					break
+				}
+
+				nextID = <-ids
+
+				// If nextID is zero value, the channel has been closed and there are no more IDs to be read.
+				if nextID == "" {
+					done = true
+					break
+				}
+
+				downloadWG.Add(1)
+				go func(id string) {
+					defer downloadWG.Done()
+					err := downloadBinary(ctx, c.dataClient, dst, id)
+					if err != nil {
+						errs <- err
+						cancel()
+						done = true
+					}
+					numFilesDownloaded.Add(1)
+				}(nextID)
+			}
+			downloadWG.Wait()
+			if numFilesDownloaded.Load()%logEveryN == 0 {
+				fmt.Fprintf(c.c.App.Writer, "downloaded %d files\n", numFilesDownloaded.Load())
+			}
+			if done {
 				break
 			}
 		}
-		if err != nil {
-			return errors.Wrapf(err, "received error from server")
+		if numFilesDownloaded.Load()%logEveryN != 0 {
+			fmt.Fprintf(c.c.App.Writer, "downloaded %d files to %s\n", numFilesDownloaded.Load(), dst)
 		}
-		data := resp.GetData()
-		last = resp.GetLast()
+	}()
+	wg.Wait()
+	close(errs)
 
-		// If no data is returned, there is no more data.
-		if len(data) == 0 {
-			break
-		}
+	if err := <-errs; err != nil {
+		return err
+	}
 
-		if len(data) != 1 {
-			return errors.Errorf("expected a single response, received %d", len(data))
-		}
+	return nil
+}
 
-		datum := data[0]
-		mdJSONBytes, err := protojson.Marshal(datum.GetMetadata())
-		if err != nil {
+// getMatchingIDs queries client for all BinaryData matching filter, and passes each of their ids into ids.
+func getMatchingBinaryIDs(ctx context.Context, client datapb.DataServiceClient, filter *datapb.Filter,
+	ids chan string, limit int,
+) error {
+	var last string
+	defer close(ids)
+	for {
+		if err := ctx.Err(); err != nil {
 			return err
 		}
 
-		timeRequested := datum.GetMetadata().GetTimeRequested().AsTime().Format(time.RFC3339)
-		fileName := timeRequested + "_" + datum.GetMetadata().GetId()
-
-		//nolint:gosec
-		jsonFile, err := os.Create(filepath.Join(dst, metadataDir, fileName+".json"))
+		resp, err := client.BinaryDataByFilter(ctx, &datapb.BinaryDataByFilterRequest{
+			DataRequest: &datapb.DataRequest{
+				Filter: filter,
+				Limit:  uint64(limit),
+				Last:   last,
+			},
+			CountOnly: false,
```

Comment:
Can also explicitly set IncludeBinary to false here.

Reply:
Done